documentation/adr/0008-pace-syntax-guiding-principles.md

```diff
@@ -0,0 +1,44 @@
+[<-previous](0007-developer-scripts-storage-location.md) | next->
+
+# 8. PACE Syntax Guiding Principles
+
+Date: 2020-Oct-19
+
+## Status
+
+Proposed
+
+
+## Context
+
+Whilst the core framework of PACE is the Horace (and Herbert) projects,
+PACE involves many other codes, including [Euphonic](https://github.com/pace-neutrons/Euphonic),
+[Brille](https://github.com/brille/brille) and [SpinW](https://spinw.org).
+The way in which these programs interact with each other is presently unclear.
+Furthermore the programs are written in several different languages.
+In this ADR we propose some guiding principles in the design of the interfaces
+between the programs or components which make up PACE.
+
+## Decision
+
+The guiding principle for PACE-supported programs is **commonality**.
+That is, interacting with the different programs of PACE should be **seamless**
+to the user, and the user interface presented to them should be **uniform**.
+
+## Consequences
+
+There are several implications stemming from this guiding principle:
+
+* There should be a single downloadble distribution which bundles all the 
+necessary components (e.g. Horace, Euphonic, Brille and SpinW at the moment).
+* User syntax should follow the same format for each program,
+and should be similar between Matlab and Python interfaces.
+Specifically the same method and keyword names should be used.
+* Syntax should be "Pythonic"(prefer keyword to positional arguments;
+reasonable defaults for as many input arguments as practicable);
+This applies to both Matlab and Python interfaces,
```

Comment:
Do you mean that _all_ positional and keyword arguments would _always_ have to be specified? This would severely restrict the flexibility of the syntax. I would definitely not want to impose this on the Python side. 

What about if we ban non-keyword string arguments? (e.g. the current `'-nopix'` would change to `'pix', false`)

(Or if non-keyword string arguments always have to be prefixed with `'-'` which would still allow `'-nopix'`)

Reply:
Sorry, didn't mean that. At the moment the MATLAB input parser offers compulsory positional, optional positional and keyword. Roughly corresponding to Python compulsory positional, *args and **kwargs. Keyword arguments are always optional. I am suggesting not having any *args, to keep things simple. If we had a way of validating optional positional string arguments to distinguish them from the first keyword, that could be possible (as you say, e.g. -nopix), but maybe for simplicity have all arguments keyword?

Reply:
Ah right. Sorry. That sounds fine then. We can omit the optional positional arguments `*args`.

Reply:
Actually, realising that I'm getting my Python syntax wrong. But it does seem that Python and MATLAB have very different ways of processing optional arguments, (and also keyword arguments: MATLAB has them as a separate category while Python has them as an alternative way of accessing the same compulsory argument set - do I have that right?) - so a minimal set of ways that work for both would maybe be the way forward. Anyway, we can discuss at the next meeting.

Reply:
So I think the compulsary/optional and positional/keyword arguments are not really part of the Matlab language as such, they're only part of the `inputParser` function. The Matlab language only really defines `varargin` which is a cell array and is equivalent to Python's `*args`. Matlab simulates Python's `**kwargs` using string-value pairs in `varargin`, so then it could have problems with string position arguments. Python, on the other hand has both positional and keyword arguments as part of the language itself. 

You're right that in Python you can access a named positional argument using a keyword with the same name as you give the positional argument variable. E.g. `def fun(a, b, c)` can be called with `fun(1, 2, 3)` _or_ `fun(a=1, b=2, c=3)` or `fun(1, 2, c=3)` with the convention that positional arguments must come first (so `fun(a=1, 2, 3)` will give a `SyntaxError`). Matlab does not support this in `inputParser` but I think if we write our own parser we could implement this if we wanted as long as we pre-define the names of the position arguments for the parser rather than just use the names in the function definition.

So, as I see it we can do one of two things in Matlab:

1. In the function definition we accept only `varargin` - that is, `function [x,y,z] = somefun(varargin)`. This cell array is passed to a parser which contains a list of *named* arguments *in order* with flags on whether they are keywords or positional, e.g. something like `[struct('name', 'arg1', 'kind', 'positional', 'type', 'numeric', 'default', 0), struct('name', 'arg2', 'kind', 'keyword', 'type', 'numeric', 'default', 0)]`. This way the user can call either `somefun('arg1', 1, 'arg2', 2)` *or* `somefun(1, 'arg2', 2)` like in Python. In this case I think we could allow the use of optional positional arguments unless the positional argument is exactly the same as a keyword argument name (we'd follow the Python convention that after the first keyword argument, every other argument is a keyword argument). 

2. Use a mix of defined (mandatory) positional arguments and `varargin` - that is `function [x,y,z] = somefun(arg1, arg2, varargin)`. In this case `varargin` contains *only* keyword arguments (its length must be even), and we disallow optional positional arguments. So the user would have to use `somefun(1, 2, 'arg3', 3)`.